feat: add chat command bubbles for received txs

jrainville opened the pull request:
Adds the chat command support for received Transactions, meaning transactions you receive from the mobile app. You can from them share your address and decline.

Support to send chat commands from the desktop app will come in another PR, along with being able to Accept sending when the other requests a tx, since this PR is already pretty huge.

Part of the hugeness is because I added the code from the BigInteger.js library. That way, we can do BigInt calculations from the JS code (https://github.com/peterolson/BigInteger.js).

Branched off of #785 

A couple of random examples:
![image](https://user-images.githubusercontent.com/11926403/92026702-5eec1f00-ed2f-11ea-9968-d9cf10310f00.png)

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 6294c125 | [#1](https://ci.status.im/job/nim-status-client/job/macos/job/PR-812/1/) | 2020-09-02 19:24:56 | ~7 min | `macos` | [:package:`dmg`](https://status-im-prs.ams3.digitaloceanspaces.com/StatusIm-200902-191723-6294c1-pr.dmg) |

richard-ramos commented on ui/shared/RecipientSelector.qml:
```diff
@@ -13,12 +13,12 @@ Item {
     //% "Invalid ethereum address"
     readonly property string addressValidationError: qsTrId("invalid-ethereum-address")
-    property bool isValid: false
+    property bool isValid: readOnly ? true : false
```
```suggestion
    property bool isValid: readOnly
```

iurimatias reviewed (changes requested):
there seems to be an issue, on the mobile I tested this with 0.01 ETH and on desktop it says 0.1ETH

emizzle commented on ui/imports/Utils.qml:
```diff
@@ -2,8 +2,14 @@ pragma Singleton
 
 import QtQuick 2.13
 import "../shared/xss.js" as XSS
+import "./BigNumber/bignumber.js" as BigNumber
```
Maybe it might be a good idea to create an npm-inited folder with a `package.json` instead of including a copy of the lib? Might need some build love though @michaelsbradleyjr thoughts?

emizzle commented on src/signals/messages.nim:
```diff
@@ -198,6 +198,19 @@ proc toMessage*(jsonMsg: JsonNode): Message =
+    .join(" ")
+
+  if message.contentType == ContentType.Transaction:
+    message.commandParameters = CommandParameters(
+      id: jsonMsg["commandParameters"]["id"].getStr,
+      fromAddress: jsonMsg["commandParameters"]["from"].getStr,
+      address: jsonMsg["commandParameters"]["address"].getStr,
+      contract: jsonMsg["commandParameters"]["contract"].getStr,
+      value: jsonMsg["commandParameters"]["value"].getStr,
+      transactionHash: jsonMsg["commandParameters"]["transactionHash"].getStr,
+      commandState: jsonMsg["commandParameters"]["commandState"].getInt,
+      signature: jsonMsg["commandParameters"]["signature"].getStr
+    )
```
Could try decoding the object instead:
```
import json_serialization

message.commandParameters = Json.decode($jsonMsg["commandParameters"], CommandParameters)
```

emizzle reviewed (changes requested):
Nice work, huge effort! I left a detailed review of the changes in the other branch that were merged in to this branch, and I'm a bit too lazy to duplicate them here 😂 : https://github.com/status-im/nim-status-client/pull/822#pullrequestreview-482323239.

I've left it as changes requested due to the comments in the link above.

richard-ramos commented on ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/ChatCommandButton.qml:
```diff
@@ -61,6 +61,8 @@ Rectangle {
         anchors.fill: parent
         cursorShape: Qt.PointingHandCursor
-        onClicked: root.onClicked
+        onClicked: {
```
just being curious,  is there a difference between using `root.onClicked` directly and putting it inside a `{}` block?

jrainville replied:
I'm not 100% to be honest. Sometimes it doesn't work well. I assume here I just changed it because I had put a `console.log` and removed it. Oh well.

jrainville commented:
I rebased and fixed the problems
@emizzle I addressed most of your comments from the old PR. The only one I didn't fix is the `estimateGas` because I'm not exactly sure what to do and you said you'd simplify it.

@iurimatias I fixed the decimal problem. The issue was with the BigInt lib I used. TLDR: it sucked. So I created custom functions that do what we need, i.e. moving the dot left or right the number of decimals the token has.

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/ChatComponents/ChatCommandModal.qml:
```diff
@@ -0,0 +1,195 @@
+                reset: function() {
+                    selectedAccount = Qt.binding(function() { return selectFromAccount.selectedAccount })
+                }
+            }
```
Adding a `GasValidator` component here would ensure the user gets feedback when they do not have enough ETH for gas (there is logic in that component that checks if they are sending ETH, then adds the tx ETH to the gas ETH). Also, adding an invisible `GasSelector` component aids in ensuring we have correct gas values (and could be made visible in the future to align with the design). This was done in the StickerPack purchase modal:

https://github.com/status-im/nim-status-client/blob/dff022bac088cd3f6f89e890e3c712c11fe33f29/ui/app/AppLayouts/Chat/components/StickerPackPurchaseModal.qml#L102-L138

However, please use the `gasEstimate` recommendation in the already merged PR (https://github.com/status-im/nim-status-client/pull/822#discussion_r485583885)

jrainville replied:
I won't add it here since it's just the request for a transaction. There is no signing any time soon, so I assume that if the user has enough Token to send to the other, he probably has enough for gas. Anyway, he can acquire it in the meantime

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/MessageComponents/TransactionComponents/SendTransactionButton.qml:
```diff
@@ -19,19 +20,51 @@ Item {
             cursorShape: Qt.PointingHandCursor
             onClicked: {
-                console.log('Sign')
+                walletModel.setFocusedAccountByAddress(commandParametersObject.fromAddress)
+                var acc = walletModel.focusedAccount
```
Why do we need to set a focused account? Is there possibly a way to simply retrieve the account by address from nim without needing to track an account?

jrainville replied:
It was possible, up until the point where I needed the `assets` of that account. There is currently no way to get a `QVariant` return type with a parameter. So I couldn't get the assets by `address`. It's kinda sucky

emizzle commented on ui/app/AppLayouts/Wallet/SendModal.qml:
```diff
@@ -14,6 +14,12 @@ ModalPopup {
     title: qsTrId("command-button-send")
     height: 504
 
+    property var selectedRecipient
+    onSelectedRecipientChanged: {
+        selectRecipient.selectedRecipient = this.selectedRecipient
+        selectRecipient.readOnly = !!this.selectedRecipient && !!this.selectedRecipient.address
+    }
```
What do you think about making the `selectedRecipient` property an alias of the `RecipientSelector` component? And any affected properties can be bound directly to its own properties. This is what I mean haha:
```suggestion
    property alias selectedRecipient: selectRecipient.selectedRecipient
    RecipientSelector {
        id: selectRecipient
        readOnly = !!selectedRecipient && !!selectedRecipient.address
    }
```

emizzle commented on src/status/signals/messages.nim:
```diff
@@ -198,6 +198,19 @@ proc toMessage*(jsonMsg: JsonNode): Message =
     message.stickerHash = jsonMsg["sticker"]["hash"].getStr
+    .join(" ")
+
+  if message.contentType == ContentType.Transaction:
+    message.commandParameters = CommandParameters(
+      id: jsonMsg["commandParameters"]["id"].getStr,
+      fromAddress: jsonMsg["commandParameters"]["from"].getStr,
+      address: jsonMsg["commandParameters"]["address"].getStr,
+      contract: jsonMsg["commandParameters"]["contract"].getStr,
+      value: jsonMsg["commandParameters"]["value"].getStr,
+      transactionHash: jsonMsg["commandParameters"]["transactionHash"].getStr,
+      commandState: jsonMsg["commandParameters"]["commandState"].getInt,
+      signature: jsonMsg["commandParameters"]["signature"].getStr
+    )
```
Could try decoding the object instead:
```
import json_serialization

message.commandParameters = Json.decode($jsonMsg["commandParameters"], CommandParameters)
```

jrainville replied:
I tried and I kept getting an error. I'm not sure why. I even changed `fromAddress` to `from` with the backticks to make it the same name, but it still doesn't work

emizzle commented on ui/app/AppLayouts/Chat/ChatColumn/MessageComponents/TransactionBubble.qml:
```diff
@@ -2,123 +2,204 @@ import QtQuick 2.3
+            return "0"
         }
+        try {
+            return Utils.divideByDecimals(commandParametersObject.value, token.decimals)
```
Instead of doing this conversion on the QML side (ie string math haha), what do you think about converting the `commandParameters.value` on the nim side with `libstatus/utils/wei2eth`?

jrainville replied:
Are you telling me we had these functions during all that time... 😭

emizzle commented on src/app/wallet/view.nim:
```diff
@@ -142,6 +145,26 @@ QtObject:
     write = setCurrentAccountByIndex
     notify = currentAccountChanged
 
+  proc focusedAccountChanged*(self: WalletView) {.signal.}
+
+  proc setFocusedAccountByAddress*(self: WalletView, address: string) {.slot.} =
+    if(self.accounts.rowCount() == 0): return
+
+    let index = self.accounts.getAccountindexByAddress(address)
+    if index == -1: return
+    let selectedAccount = self.accounts.getAccount(index)
+    if self.focusedAccount.address == selectedAccount.address: return
+    self.focusedAccount.setAccountItem(selectedAccount)
+    self.focusedAccountChanged()
+
+  proc getFocusedAccount*(self: WalletView): QVariant {.slot.} =
+    result = newQVariant(self.focusedAccount)
+
+  QtProperty[QVariant] focusedAccount:
+    read = getFocusedAccount
+    write = setFocusedAccountByAddress
+    notify = focusedAccountChanged
+
```
What do you think about replacing all of this with an account lookup by address? See other comment on the QML side.

emizzle reviewed (commented):
Well done again! I added a few more suggestions. I also left a few of the same review comments that were outdated, that maybe you saw but decided not to implement, just in case they potentially got lost in the sea of multiple PRs.

jrainville commented:
@emizzle thanks again for the great comments. I think I updated in regards to each of them this time. Some of them I didn't do just because it didn't work (eg: serialization)

iurimatias commented:
nice
